src/net/ossl.cc

```diff
@@ -453,5 +462,592 @@ void tls::server_credentials::set_client_auth(client_auth ca) {
+        // read from input until we see EOF. Any other reader
+        // before us will get it instead of us, and mark _eof = true
+        // in which case we will be no-op.
+        return with_semaphore(_out_sem, 1,
+                        std::bind(&session::do_shutdown, this)).then(
+                        std::bind(&session::wait_for_eof, this)).finally([me = shared_from_this()] {});
+        // note moved finally clause above. It is theorethically possible
+        // that we could complete do_shutdown just before the close calls
+        // below, get pre-empted, have "close()" finish, get freed, and
+        // then call wait_for_eof on stale pointer.
+    }
+    void close() noexcept {
+        // only do once.
+        if (!std::exchange(_shutdown, true)) {
+            netlogger.info("Invoking shutdown routine within session::close");
+            auto me = shared_from_this();
+            // running in background. try to bye-handshake us nicely, but after 10s we forcefully close.
+            (void)with_timeout(timer<>::clock::now() + std::chrono::seconds(10), shutdown()).finally([this] {
+                _eof = true;
+                try {
+                    (void)_in.close().handle_exception([](std::exception_ptr) {}); // should wake any waiters
+                } catch (...) {
+                }
+                try {
+                    (void)_out.close().handle_exception([](std::exception_ptr) {});
+                } catch (...) {
+                }
+                // make sure to wait for handshake attempt to leave semaphores. Must be in same order as
+                // handshake aqcuire, because in worst case, we get here while a reader is attempting
+                // re-handshake.
+                return with_semaphore(_in_sem, 1, [this] {
+                    return with_semaphore(_out_sem, 1, [] {});
+                });
+            }).then_wrapped([me = std::move(me)](future<> f) { // must keep object alive until here.
+                f.ignore_ready_future();
+            });
+        }
+    }
+    // helper for sink
+    future<> flush() noexcept {
+        return with_semaphore(_out_sem, 1, [this] {
+            return _out.flush();
+        });
+    }
+
+    seastar::net::connected_socket_impl & socket() const {
+        return *_sock;
+    }
+
+    /// TODO(rob) - implement these methods
+    future<std::optional<session_dn>> get_distinguished_name() {
+        using result_t = std::optional<session_dn>;
+        return make_exception_future<result_t>(std::runtime_error("unimplemented"));
+    }
+
+    future<std::vector<subject_alt_name>> get_alt_name_information(std::unordered_set<subject_alt_name_type>) {
+        using result_t = std::vector<subject_alt_name>;
+        return make_exception_future<result_t>(std::runtime_error("unimplemented"));
+    }
+
+
+private:
+    std::optional<session_dn> extract_dn_information() const {
+        auto peer_cert = x509_ptr(SSL_get_peer_certificate(_ssl.get()));
```

Comment:
Suggest using either `SSL_get0_peer_certificate` so you don't need to manage the lifetime of the cert.